From 5d2e16e1350198e3fdefb6a5afc1edbd3581453e Mon Sep 17 00:00:00 2001
From: Sarath Babu Naidu Gaddam <sarath.babu.naidu.gaddam@amd.com>
Date: Mon, 28 Nov 2022 11:08:15 +0530
Subject: [LINUX PATCH 2/2] net: xilinx: Add QPLL reset for XXV MAC

Reference clock input to XXV subsystem may be initialized during
linux probe (with manufacturer specific clock programming drivers).
This clock may not be ready when the FPGA region is loaded with XXV
subsystem design. Hence perform a QPLL reset at the beginning of XXV
Ethernet initialization to ensure link up. This property is optional,
dependent on the HW system design.

Signed-off-by: Sarath Babu Naidu Gaddam <sarath.babu.naidu.gaddam@amd.com>
---
 drivers/net/ethernet/xilinx/xilinx_axienet.h  |  2 ++
 .../net/ethernet/xilinx/xilinx_axienet_main.c | 25 +++++++++++++++++++
 2 files changed, 27 insertions(+)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet.h b/drivers/net/ethernet/xilinx/xilinx_axienet.h
index e2f3c27d45d7..ac12c31042bf 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet.h
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet.h
@@ -751,6 +751,7 @@ struct aximcdma_bd {
  * @gt_lane: MRMAC GT lane index used.
  * @ptp_os_cf: CF TS of PTP PDelay req for one step usage.
  * @xxv_ip_version: XXV IP version
+ * @reset_gpio: Reset the QPLL for XXV
  */
 struct axienet_local {
 	struct net_device *ndev;
@@ -831,6 +832,7 @@ struct axienet_local {
 	u32 gt_lane;		/* MRMAC GT lane index used */
 	u64 ptp_os_cf;		/* CF TS of PTP PDelay req for one step usage */
 	u32 xxv_ip_version;
+	struct gpio_desc *reset_gpio; /* Reset the QPLL for XXV */
 };
 
 /**
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index da69f890acfc..46a82fd24f7b 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -46,6 +46,7 @@
 #include <linux/xilinx_phy.h>
 #include <linux/clk.h>
 #include <linux/ptp/ptp_xilinx.h>
+#include <linux/gpio/consumer.h>
 
 #include "xilinx_axienet.h"
 
@@ -552,6 +553,21 @@ void __axienet_device_reset(struct axienet_dma_q *q)
 	}
 }
 
+/**
+ *axienet_gpio_reset_qpll - Reset QPLL
+ *@lp Pointer to the axienet_local structure
+ *
+ *This function is called to reset the QPLL for XXV MAC.This reset is used
+ * when IDT clock programming is done by Linux clock matric driver.
+ */
+
+static void axienet_gpio_reset_qpll(struct axienet_local *lp)
+{
+	gpiod_set_value(lp->reset_gpio, 1);
+	msleep(5);
+	gpiod_set_value(lp->reset_gpio, 0);
+}
+
 /**
  * axienet_device_reset - Reset and initialize the Axi Ethernet hardware.
  * @ndev:	Pointer to the net_device structure
@@ -576,6 +592,8 @@ static int axienet_device_reset(struct net_device *ndev)
 	u8 maj, minor;
 
 	if (lp->axienet_config->mactype == XAXIENET_10G_25G) {
+		/* Reset the QPLL */
+		axienet_gpio_reset_qpll(lp);
 		/* Reset the XXV MAC */
 		val = axienet_ior(lp, XXV_GT_RESET_OFFSET);
 		val |= XXV_GT_RESET_MASK;
@@ -3497,6 +3515,13 @@ static int axienet_probe(struct platform_device *pdev)
 			goto free_netdev;
 		}
 
+		if (lp->axienet_config->mactype == XAXIENET_10G_25G) {
+			lp->reset_gpio = devm_gpiod_get_optional(lp->dev, "reset", GPIOD_OUT_LOW);
+			if (IS_ERR(lp->reset_gpio)) {
+				dev_warn(&pdev->dev, "QPLL GPIO reset property not found\n");
+			}
+		}
+
 		if (dma_set_mask_and_coherent(lp->dev, DMA_BIT_MASK(lp->dma_mask)) != 0) {
 			dev_warn(&pdev->dev, "default to %d-bit dma mask\n", XAE_DMA_MASK_MIN);
 			if (dma_set_mask_and_coherent(lp->dev,
-- 
2.25.1

